core/includes/config.inc

```diff
@@ -1232,6 +1241,345 @@ interface ConfigStorageInterface {
   public function exportArchive($file_uri);
 }
 
+/**
+ * Defines the database storage controller.
+ */
+class ConfigDatabaseStorage implements ConfigStorageInterface {
+
+  /**
+   * The database table to use for configuration objects.
+   *
+   * @var string
+   */
+  protected $table = '';
+
+  /**
+   * The database connection to use.
+   *
+   * @var string
+   */
+  protected $database = '';
+
+  /**
+   * The database schema.
+   */
+  protected function schema($name) {
+    $schema = array(
+      'description' => 'Stores the configuration of the site in database tables.',
+      'fields' => array(
+        'name' => array(
+          'type' => 'varchar',
+          'length' => 255,
+          'not null' => TRUE,
+          'description' => 'The top-level name for the full key.  Would be the filename in the ConfigFileStore, minus the .json extension',
+        ),
+        'data' => array(
+          'type' => 'text',
+          'size' => 'big',
+          'not null' => TRUE,
+          'description' => 'The JSON encoded value for the key.',
+        ),
+        'ctime' => array(
+          'type' => 'int',
+          'not null' => TRUE,
+          'default' => 0,
+          'description' => 'The node changed time (see stat(2) for details.)',
+        ),
+      ),
+      'primary key' => array('name'),
+    );
+
+    return $schema;
+  }
+
+  /**
+   * Constructs a new ConfigDatabaseStorage controller.
+   *
+   * @param string $dburl
+   *   A URL that references a backdrop connection (optional) and table. The
+   *   string has a format of db:/<database connection>/<database table>.
+   *
+   *   Examples:
+   *    * db:/default/config_active
+   *    * db:config_active
+   *    * db://config_active
+   *    * db:/config_store/config_stage
+   *
+   *   If no database connection is specified, then the 'default' connection is
+   *   used.
+   */
+  public function __construct($dburl) {
+    $matches = array();
+    preg_match('/^db:(\/(\w*)\/)?(\w+)$/', trim($dburl), $matches);
+    if (count($matches) != 4) {
+      throw new ConfigStorageException(t('Invalid database specifier: @db', array('@db' => $dburl)));
+    }
+    $this->table = $matches[3];
+    $this->database = $matches[2] ? $matches[2] : 'default';
+
+    // Bootstrap the database if it is not yet available.
+    if (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE) {
+      require_once BACKDROP_ROOT . '/core/includes/database/database.inc';
```

Comment:
It should call `backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE)`, since [`_backdrop_bootstrap_database()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21bootstrap.inc/function/_backdrop_bootstrap_database/1) does more than just requiring a file. It also changes the database prefixes when tests are running.

Reply:
@laryn there's probably some duplication. I agree that we should add `backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE)` in addition to including the database.inc file. But then we probably won't need this in bootstrap.inc:

```
$config_storage = config_get_config_storage('active');
  if (is_a($config_storage, 'ConfigDatabaseStorage') && (!function_exists('db_query') || backdrop_get_bootstrap_phase() < BACKDROP_BOOTSTRAP_DATABASE)) {
    // backdrop_bootstrap(BACKDROP_BOOTSTRAP_DATABASE, FALSE);
  }
```

because calling `config_get_config_storage()` will initialize the `ConfigDatabaseStorage` class which will bootstrap the database.

Reply:
I've confirmed my previous comment. Call the function here and remove it from bootstrap.inc and it installs fine.